___

<!-- <a><img src="https://colab.research.google.com/assets/colab-badge.svg" align="left"></a>

<img src="https://raw.githubusercontent.com/jupyter/design/master/logos/Badges/nbviewer_badge.svg" align="right"><br> -->


<a href=""><img align="left" src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open in Colab" title="This will be active in Solution Notebook"></a>

<a href="" style="float: right" target="_blank">
    <img src="https://raw.githubusercontent.com/jupyter/design/master/logos/Badges/nbviewer_badge.svg" alt="Open in nbbviewer" title="This will be active in Solution Notebook">
</a><br>

<a href=""><img align="center" width="80" height="80" src="https://i.ibb.co/Vqpq8gQ/Git-Hub-logo.png" alt="Open in Colab" title="This will be active in Solution Notebook"></a>

___

<a href="https://clarusway.com/"><img align="center" src="https://i.ibb.co/Rpz9L36/clarusway-logo-black.png" width="450" height="200" alt="CLRSWY_LOGO" title="Open and Execute in Google Colaboratory"></a><br/>

<p style="background-color:#E51A59; font-family:newtimeroman; color:#FDFEFE; font-size:130%; text-align:center; border-radius:10px 10px;">WAY TO REINVENT YOURSELF</p>

<p style="text-align: center;"><img src="https://i.ibb.co/99DQ3TY/DS-Courses.png" class="img-fluid" 
alt="CLRSWY"></p>

## <p style="background-color:#FDFEFE; font-family:newtimeroman; color:#060108; font-size:200%; text-align:center; border-radius:10px 10px;">Data Analysis with Python</p>

![Image_Assignment](https://i.ibb.co/cY9HhF7/bnote-assignment.gif)

## <p style="background-color:#FDFEFE; font-family:newtimeroman; color:#4d77cf; font-size:200%; text-align:center; border-radius:10px 10px;">Assignment 02 (Pandas)</p>

<a id="toc"></a>

## <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">CONTENT</p>


* [CONTEXT](#0)
* [ABOUT THE DATASET & ATRIBUTES](#00)
* [INSTRUCTIONS (Tasks in the Assignment)](#000)
* [IMPORTING LIBRARIES NEEDED IN THIS NOTEBOOK](#1)
* [Task 01. Checking the head of the DataFrame](#2)
* [Task 02. How many rows and columns are there?](#3)
* [Task 03. What is the average Purchase Price?](#4)
* [Task 04. What were the highest and lowest purchase prices?](#5) 
* [Task 05. Listing the country names and number of records belongs to each country](#6)
* [Task 06. How many records are as Country United Kingdom?](#7)
* [Task 07. Creating a column named Total which shows the total amount of spend using UnitPrice and Quantity](#8)
* [Task 08. Showing total spending amount according to the Countries](#9)
* [Task 09. Which Country spend least?](#10)
* [Task 10. Calculating average of spending according to country with apply function](#11) 
* [Task 11. Calculating average, mean, max of spending according to country with agg function](#12)
* [Task 12. What is the max Total spending in whole dataset and show the records of that spending?](#13)
* [Task 13. How many people who live in Spain spend more than average Total?](#14)
* [Task 14. Converting datatype of InvoiceDate to datetime and assign it to new column name InvoiceDate2](#15)
* [Task 15. Taking only date values (year-month-day) from InvoiceDate2 and assign it to a new column named "date"](#16)
* [Task 16. Dropping InvoiceDate2 and InvoiceDate columns permanently](#17)
* [Task 17. Removing A, B, G and E characters at the first 5 rows of "StockCode" in ecom dataframe with a string method](#18) 
* [Task 18. Taking only Total and date columns with loc and craete a new df named ecom2](#19)
* [Task 19. Taking only InvoiceNo and CustomerID columns with iloc and craete a new df named ecom3](#20)
* [Task 20. Combining ecom2 and ecom3 with concat() method](#21)
* [Task 21. Combining the common (intersected) values of ecom2 and ecom3 with merge() method](#22)
* [Task 22. Combining ecom2 and ecom3 with join() method](#23)
* [THE END OF THE ASSIGNMENT 02](#24)

<a id="toc"></a>

## <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">CONTEXT</p>

<a id="0"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

E-Commerce is currently one of the fastest and dynamically evolving industries in the world. Its popularity has been growing rapidly with the ease of digital transactions and quick door-to-door deliveries. A major source of big tech companies revenues comes from the interaction of their underlying proprietary algorithms that are heavily powered by data science, so it’s fundamental to understand the methods applied to maintain and grow the number of clients.

"This is a transnational data set which contains all the transactions occurring between 01/12/2010 and 09/12/2011 for a UK-based and registered non-store online retail.The company mainly sells unique all-occasion gifts. Many customers of the company are wholesalers."

<a id="toc"></a>

## <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">ABOUT THE DATASET & ATRIBUTES</p>

<a id="00"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

Typically e-commerce datasets are proprietary and consequently hard to find among publicly available data. However, The [UCI Machine Learning Repository](http://archive.ics.uci.edu/ml/index.php) has made this dataset containing actual transactions from 2010 and 2011. The dataset is maintained on their site, where it can be found by the title "Online Retail".

**Attribute Information:**

**``InvoiceNo:``** Invoice number. Nominal, a 6-digit integral number uniquely assigned to each transaction. If this code starts with letter 'c', it indicates a cancellation.<br>
**``StockCode:``** Product (item) code. Nominal, a 5-digit integral number uniquely assigned to each distinct product.<br>
**``Description:``** Product (item) name. Nominal.<br>
**``Quantity:``** The quantities of each product (item) per transaction. Numeric.<br>
**``InvoiceDate:``** Invice Date and time. Numeric, the day and time when each transaction was generated.<br>
**``UnitPrice:``** Unit price. Numeric, Product price per unit in sterling.<br>
**``CustomerID:``** Customer number. Nominal, a 5-digit integral number uniquely assigned to each customer.<br>
**``Country:``** Country name. Nominal, the name of the country where each customer resides.<br>

## <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">INSTRUCTIONS (Tasks in the Assignment)</p>

<a id="000"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

**In this assignment you're required to handle with the following tasks:** 

- **Task 01.** Checking the head of the DataFrame.<br>
- **Task 02.** How many rows and columns are there?<br>
- **Task 03.** What is the average of Product sold and the average price of Product per unit?<br>
- **Task 04.** What were the highest and lowest purchase prices?<br>
- **Task 05.** Listing the country names and number of their records belongs to each country.<br>
- **Task 06.** How many records are as Country United Kingdom?<br>
- **Task 07.** Creating a column named "Total" which shows the total amount of spend using UnitPrice and Quantity.<br>
- **Task 08.** Showing total spending amount according to the Countries and their percentages.<br>
- **Task 09.** Which Country spend least?<br>
- **Task 10.** Calculating average of spending according to country with apply function  and create a new feature named "Average_Spend".<br>
- **Task 11.** Calculating average, min, max of spending according to country with agg function and create new features named "Max_Spend" and "Min_Spend".<br>
- **Task 12.** What is the max Total spending in whole dataset and show all records of that spending?<br>
- **Task 13.** How many people who live in Spain spend more than average Total?<br>
- **Task 14.** Converting datatype of InvoiceDate to datetime and assign it to new column name InvoiceDate2.<br>
- **Task 15.** Taking only date values (year-month-day) from InvoiceDate2 and assign it to a new column named "date".<br>
- **Task 16.** Dropping InvoiceDate2 and InvoiceDate columns permanently.<br>
- **Task 17.** Removing A, B, G and E characters at the first 5 rows of "StockCode" in ecom dataframe with a string method.<br>
- **Task 18.** Taking only Total and date columns with loc and craete a new df named ecom2.<br>
- **Task 19.** Taking only InvoiceNo and CustomerID columns with iloc and craete a new df named ecom3.<br>
- **Task 20.** Combining ecom2 and ecom3 with concat() method.<br>
- **Task 21.** Combining the common (intersected) values of ecom2 and ecom3 with merge() method.<br>
- **Task 22.** Combining ecom2 and ecom3 with join() method.<br>

## <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">IMPORTING LIBRARIES NEEDED IN THIS STUDY</p>

<a id="1"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

Once you've installed NumPy and Pandas you can import them as a library. Then read the dataset and assig it as **``"ecom"``**:

In [5]:
# Import the related library to this task

# YOUR CODE IS HERE 
import pandas as pd


In [6]:
# After reading the dataset, assign it named "ecom"

# YOUR CODE IS HERE 

ecom = pd.read_csv("data.csv", encoding='latin-1')

FileNotFoundError: [Errno 2] No such file or directory: 'data.csv'

### 🔥 SPECIAL NOTE! 🔥	

**If "MemoryError" occurs, take only the first 100 rows or some part of DataFrame**

**[Not reading all rows while importing csv into pandas dataframe](https://stackoverflow.com/questions/33161769/not-reading-all-rows-while-importing-csv-into-pandas-dataframe)**

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 01: Check the Head of the DataFrame</p>

<a id="2"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [ ]:
# YOUR CODE IS HERE 

ecom.head(10)

Desired Output:

![image.png](https://i.ibb.co/LdKJHb6/assignment1a.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 02: How many rows and columns are there?</p>

<a id="3"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [4]:
# YOUR CODE IS HERE 

ecom.info()

NameError: name 'ecom' is not defined

In [10]:
# YOUR CODE IS HERE
ecom.shape


(541909, 8)

### Desired Output:

(541909, 8)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 03: What is the average of Product sold and the average price of Product per unit?</p>

<a id="4"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [11]:
# YOUR CODE IS HERE 
ecom['UnitPrice'].mean()

4.611113626088513

In [12]:
# YOUR CODE IS HERE 

ecom['Quantity'].mean()

9.55224954743324

In [13]:
# YOUR CODE IS HERE 

ecom[['UnitPrice', 'Quantity']].mean()

UnitPrice    4.611114
Quantity     9.552250
dtype: float64

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 04: What are the highest and lowest purchase prices?</p>

<a id="5"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [14]:
# YOUR CODE IS HERE 
ecom['UnitPrice'].max()


38970.0

In [15]:
# YOUR CODE IS HERE 
ecom['UnitPrice'].min()


-11062.06

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 05: List the country names and number of their records belongs to each country</p>

<a id="6"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [16]:
# YOUR CODE IS HERE 
ecom.groupby('Country').size().sort_values(ascending=False)

Country
United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


In [1]:
ecom.Country.value_counts(dropna=False)

NameError: name 'ecom' is not defined

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 06: How many records are there in the United Kingdom?</p>

<a id="7"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [17]:
# YOUR CODE IS HERE 

ecom[ecom['Country']=='United Kingdom'].count()

InvoiceNo      495478
StockCode      495478
Description    494024
Quantity       495478
InvoiceDate    495478
UnitPrice      495478
CustomerID     361878
Country        495478
dtype: int64

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 07: Create a column named "Total" which shows the total amount of spend using UnitPrice and Quantity</p>

<a id="8"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [18]:
# YOUR CODE IS HERE 
ecom['Total'] = ecom['UnitPrice'] * ecom['Quantity']
ecom['Total']

0         15.30
1         20.34
2         22.00
3         20.34
4         20.34
          ...  
541904    10.20
541905    12.60
541906    16.60
541907    16.60
541908    14.85
Name: Total, Length: 541909, dtype: float64

In [19]:
# YOUR CODE IS HERE 
ecom.head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365     85123   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365     84406       CREAM CUPID HEARTS COAT HANGER         8   
3    536365     84029  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365     84029       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  Total  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  15.30  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  22.00  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34

Desired Output:

![image.png](https://i.ibb.co/7y5d0sp/assignment2b.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 08: Show total spending amount according to the Countries and their percentages.</p>

<a id="9"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [20]:
# YOUR CODE IS HERE 
ecom.groupby('Country')['Total'].sum()


Country
Australia                137077.270
Austria                   10154.320
Bahrain                     548.400
Belgium                   40910.960
Brazil                     1143.600
Canada                     3666.380
Channel Islands           20086.290
Cyprus                    12946.290
Czech Republic              707.720
Denmark                   18768.140
EIRE                     263276.820
European Community         1291.750
Finland                   22326.740
France                   197403.900
Germany                  221698.210
Greece                     4710.520
Hong Kong                 10117.040
Iceland                    4310.000
Israel                     7907.820
Italy                     16890.510
Japan                     35340.620
Lebanon                    1693.880
Lithuania                  1661.060
Malta                      2505.470
Netherlands              284661.540
Norway                    35163.460
Poland                     7213.140
Portugal            

In [21]:
# YOUR CODE IS HERE 
(ecom.groupby('Country')['Total'].sum() / ecom['Total'].sum() * 100).sort_values(ascending=False)

Country
United Kingdom          83.996903
Netherlands              2.920280
EIRE                     2.700899
Germany                  2.274353
France                   2.025123
Australia                1.406246
Switzerland              0.578445
Spain                    0.561920
Belgium                  0.419697
Sweden                   0.375429
Japan                    0.362552
Norway                   0.360734
Portugal                 0.301270
Finland                  0.229045
Channel Islands          0.206061
Denmark                  0.192538
Italy                    0.173276
Cyprus                   0.132813
Austria                  0.104171
Hong Kong                0.103788
Singapore                0.093564
Israel                   0.081125
Poland                   0.073998
Unspecified              0.048727
Greece                   0.048324
Iceland                  0.044215
Canada                   0.037613
Malta                    0.025703
United Arab Emirates     0.019515
USA   

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 09: Which Country spend least?</p>

<a id="10"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [22]:
# YOUR CODE IS HERE
ecom.groupby('Country')['Total'].sum().sort_values().head(1)

Country
Saudi Arabia    131.17
Name: Total, dtype: float64

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 10: Calculate the average of spending according to country with apply function and create a new feature named "Average_Spend"</p>

<a id="11"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [23]:
# YOUR CODE IS HERE 
ecom.groupby('Country')['Total'].apply(lambda x: x.mean()) 


Country
Australia               108.877895
Austria                  25.322494
Bahrain                  28.863158
Belgium                  19.773301
Brazil                   35.737500
Canada                   24.280662
Channel Islands          26.499063
Cyprus                   20.813971
Czech Republic           23.590667
Denmark                  48.247147
EIRE                     32.122599
European Community       21.176230
Finland                  32.124806
France                   23.069288
Germany                  23.348943
Greece                   32.263836
Hong Kong                35.128611
Iceland                  23.681319
Israel                   26.625657
Italy                    21.034259
Japan                    98.716816
Lebanon                  37.641778
Lithuania                47.458857
Malta                    19.728110
Netherlands             120.059696
Norway                   32.378877
Poland                   21.152903
Portugal                 19.333127
RSA         

In [24]:
# YOUR CODE IS HERE 

ecom['Average_Spend'] = ecom.groupby('Country')['Total'].transform(lambda x: x.mean()) 
ecom.head(3)

InvoiceNo StockCode                         Description  Quantity  \
0    536365     85123  WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                 WHITE METAL LANTERN         6   
2    536365     84406      CREAM CUPID HEARTS COAT HANGER         8   

      InvoiceDate  UnitPrice  CustomerID         Country  Total  Average_Spend  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  15.30      16.525065  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34      16.525065  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  22.00      16.525065

Desired Output:

![image.png](https://i.ibb.co/b3m1jhs/assignment2c.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 11: Calculate average, min, max of spending according to country with agg() function and create new features named "Max_Spend" and "Min_Spend"</p>

<a id="12"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [25]:
# YOUR CODE IS HERE 
ecom.groupby('Country')['Total'].agg([min,'mean', max])


C:\Users\nurda\AppData\Local\Temp\ipykernel_23368\691302728.py:2: FutureWarning: The provided callable <built-in function min> is currently using SeriesGroupBy.min. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "min" instead.
  ecom.groupby('Country')['Total'].agg([min,'mean', max])
C:\Users\nurda\AppData\Local\Temp\ipykernel_23368\691302728.py:2: FutureWarning: The provided callable <built-in function max> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ecom.groupby('Country')['Total'].agg([min,'mean', max])


min        mean        max
Country                                               
Australia               -425.00  108.877895    1718.40
Austria                  -21.25   25.322494     360.00
Bahrain                 -205.74   28.863158     205.74
Belgium                  -19.95   19.773301     165.00
Brazil                    15.00   35.737500     175.20
Canada                     2.50   24.280662     550.94
Channel Islands         -293.00   26.499063     408.00
Cyprus                  -320.69   20.813971     320.69
Czech Republic           -35.76   23.590667      70.80
Denmark                  -59.80   48.247147     428.40
EIRE                   -1917.00   32.122599    2365.20
European Community        -8.50   21.176230      60.00
Finland                  -80.00   32.124806     551.20
France                 -8322.12   23.069288    4161.06
Germany                 -599.50   23.348943     876.00
Greece                   -50.00   32.263836     175.20
Hong Kong              -2653.95   35.128611    2653.95
Iceland                    5.04   23.681319     249.60
Israel                  -136.00   26.625657     350.40
Italy                    -89.55   21.034259     300.00
Japan                  -1591.20   98.716816    3794.40
Lebanon                   10.20   37.641778     153.00
Lithuania                 13.20   47.458857     122.40
Malta                   -130.00   19.728110     455.00
Netherlands             -266.40  120.059696    4992.00
Norway                  -376.50   32.378877     700.00
Poland                   -39.80   21.152903     120.00
Portugal               -1241.98   19.333127    1241.98
RSA                        0.00   17.281207      38.25
Saudi Arabia             -14.75   13.117000      19.80
Singapore              -3949.32   39.827031    3949.32
Spain                  -1715.85   21.624390    1350.00
Sweden                 -1188.00   79.211926    1188.00
Switzerland             -179.00   28.164510     360.00
USA                      -74.88    5.948179      81.36
United Arab Emirates       3.48   27.974706     153.00
United Kingdom       -168469.60   16.525065  168469.60
Unspecified                0.19   10.649753      69.36

Desired Output:

![image.png](https://i.ibb.co/0qRCRjH/assignment2d.png)

In [26]:
# YOUR CODE IS HERE 

ecom['Max_Spend'] = ecom.groupby('Country')['Total'].transform(max)
ecom['Min_Spend'] = ecom.groupby('Country')['Total'].transform(min)
ecom

C:\Users\nurda\AppData\Local\Temp\ipykernel_23368\1862624263.py:3: FutureWarning: The provided callable <built-in function max> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ecom['Max_Spend'] = ecom.groupby('Country')['Total'].transform(max)
C:\Users\nurda\AppData\Local\Temp\ipykernel_23368\1862624263.py:4: FutureWarning: The provided callable <built-in function min> is currently using SeriesGroupBy.min. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "min" instead.
  ecom['Min_Spend'] = ecom.groupby('Country')['Total'].transform(min)


InvoiceNo StockCode                          Description  Quantity  \
0         536365     85123   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365     84406       CREAM CUPID HEARTS COAT HANGER         8   
3         536365     84029  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365     84029       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  Total  \
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  15.30   
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34   
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  22.00   
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34   
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34   
...                 ...        ...         ...             ...    ...   
541904  12/9/2011 12:50       0.85     12680.0          France  10.20   
541905  12/9/2011 12:50       2.10     12680.0          France  12.60   
541906  12/9/2011 12:50       4.15     12680.0          France  16.60   
541907  12/9/2011 12:50       4.15     12680.0          France  16.60   
541908  12/9/2011 12:50       4.95     12680.0          France  14.85   

        Average_Spend  Max_Spend  Min_Spend  
0           16.525065  168469.60 -168469.60  
1           16.525065  168469.60 -168469.60  
2           16.525065  168469.60 -168469.60  
3           16.525065  168469.60 -168469.60  
4           16.525065  168469.60 -168469.60  
...               ...        ...        ...  
541904      23.069288    4161.06   -8322.12  
541905      23.069288    4161.06   -8322.12  
541906      23.069288    4161.06   -8322.12  
541907      23.069288    4161.06   -8322.12  
541908      23.069288    4161.06   -8322.12  

[541909 rows x 12 columns]

Desired Output:

![image.png](https://i.ibb.co/Smd3Tkt/assignment2e.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 12: What is the max Total spending in whole dataset and show all records of that spending?</p>

<a id="13"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [27]:
# YOUR CODE IS HERE
ecom[ecom['Total'] == ecom['Total'].max()]

InvoiceNo StockCode                  Description  Quantity  \
540421    581483     23843  PAPER CRAFT , LITTLE BIRDIE     80995   

           InvoiceDate  UnitPrice  CustomerID         Country     Total  \
540421  12/9/2011 9:15       2.08     16446.0  United Kingdom  168469.6   

        Average_Spend  Max_Spend  Min_Spend  
540421      16.525065   168469.6  -168469.6

Desired Output:

![image.png](https://i.ibb.co/2W9t4nh/assignment2f.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 13: How many people who live in Spain spend more than average Total?</p>

<a id="14"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [28]:
# YOUR CODE IS HERE 

ecom[(ecom['Country'] == 'Spain') & (ecom['Total'] > ecom['Total'].mean())].count()

InvoiceNo        728
StockCode        728
Description      728
Quantity         728
InvoiceDate      728
UnitPrice        728
CustomerID       728
Country          728
Total            728
Average_Spend    728
Max_Spend        728
Min_Spend        728
dtype: int64

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 14: Convert datatype of "InvoiceDate" to datetime and assign it to new column named "InvoiceDate2"</p>

<a id="15"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [29]:
ecom['InvoiceDate2'] = pd.to_datetime(ecom['InvoiceDate'])
ecom['InvoiceDate2']

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate2, Length: 541909, dtype: datetime64[ns]

In [30]:
# YOUR CODE IS HERE 

ecom.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 13 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   InvoiceNo      541909 non-null  object        
 1   StockCode      541909 non-null  object        
 2   Description    540455 non-null  object        
 3   Quantity       541909 non-null  int64         
 4   InvoiceDate    541909 non-null  object        
 5   UnitPrice      541909 non-null  float64       
 6   CustomerID     406829 non-null  float64       
 7   Country        541909 non-null  object        
 8   Total          541909 non-null  float64       
 9   Average_Spend  541909 non-null  float64       
 10  Max_Spend      541909 non-null  float64       
 11  Min_Spend      541909 non-null  float64       
 12  InvoiceDate2   541909 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(6), int64(1), object(5)
memory usage: 53.7+ MB


### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 15: Take only date values (year-month-day) from InvoiceDate2 and assign it to a new column named **``date``**</p>

<a id="16"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [31]:
# YOUR CODE IS HERE 
ecom['date'] = ecom['InvoiceDate2'].dt.date
ecom['date']


0         2010-12-01
1         2010-12-01
2         2010-12-01
3         2010-12-01
4         2010-12-01
             ...    
541904    2011-12-09
541905    2011-12-09
541906    2011-12-09
541907    2011-12-09
541908    2011-12-09
Name: date, Length: 541909, dtype: object

In [32]:
# YOUR CODE IS HERE 

ecom['date'] = ecom['InvoiceDate2'].dt.date
ecom.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365     85123   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365     84406       CREAM CUPID HEARTS COAT HANGER         8   
3    536365     84029  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365     84029       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  Total  \
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  15.30   
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34   
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  22.00   
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34   
4  12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34   

   Average_Spend  Max_Spend  Min_Spend        InvoiceDate2        date  
0      16.525065   168469.6  -168469.6 2010-12-01 08:26:00  2010-12-01  
1      16.525065   168469.6  -168469.6 2010-12-01 08:26:00  2010-12-01  
2      16.525065   168469.6  -168469.6 2010-12-01 08:26:00  2010-12-01  
3      16.525065   168469.6  -168469.6 2010-12-01 08:26:00  2010-12-01  
4      16.525065   168469.6  -168469.6 2010-12-01 08:26:00  2010-12-01

Desired Output:

![image.png](https://i.ibb.co/hdQm0jx/assignment2g.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 16: Drop "InvoiceDate2" and "InvoiceDate" columns permanently</p>

<a id="17"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [33]:
# YOUR CODE IS HERE 

ecom.drop(columns=['InvoiceDate2', 'InvoiceDate'], inplace=True)
ecom.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365     85123   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365     84406       CREAM CUPID HEARTS COAT HANGER         8   
3    536365     84029  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365     84029       RED WOOLLY HOTTIE WHITE HEART.         6   

   UnitPrice  CustomerID         Country  Total  Average_Spend  Max_Spend  \
0       2.55     17850.0  United Kingdom  15.30      16.525065   168469.6   
1       3.39     17850.0  United Kingdom  20.34      16.525065   168469.6   
2       2.75     17850.0  United Kingdom  22.00      16.525065   168469.6   
3       3.39     17850.0  United Kingdom  20.34      16.525065   168469.6   
4       3.39     17850.0  United Kingdom  20.34      16.525065   168469.6   

   Min_Spend        date  
0  -168469.6  2010-12-01  
1  -168469.6  2010-12-01  
2  -168469.6  2010-12-01  
3  -168469.6  2010-12-01  
4  -168469.6  2010-12-01

Desired Output:

![image.png](https://i.ibb.co/2dPtL0k/assignment2h.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 17: Remove A, B, G and E characters at the first 5 rows of "StockCode" in ecom dataframe with a string method</p>

<a id="18"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [34]:
# YOUR CODE IS HERE 

ecom.loc[:4, 'StockCode'] = ecom.loc[:4, 'StockCode'].str.strip('AGBE')
ecom.loc[:5, 'StockCode']

0    85123
1    71053
2    84406
3    84029
4    84029
5    22752
Name: StockCode, dtype: object

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 18: Take only "Total" and "date" columns with loc and craete a new df named **``ecom2``**</p>

<a id="19"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [35]:
# YOUR CODE IS HERE        
ecom2 = ecom.loc[:, ['Total', 'date']]
ecom2                                                                                                          

Total        date
0       15.30  2010-12-01
1       20.34  2010-12-01
2       22.00  2010-12-01
3       20.34  2010-12-01
4       20.34  2010-12-01
...       ...         ...
541904  10.20  2011-12-09
541905  12.60  2011-12-09
541906  16.60  2011-12-09
541907  16.60  2011-12-09
541908  14.85  2011-12-09

[541909 rows x 2 columns]

Desired Output:

![image.png](https://i.ibb.co/yFyxdCz/assignment2j.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 19: Take only "InvoiceNo" "CustomerID" and "Total" columns with iloc and craete a new df named **``ecom3``**</p>

<a id="20"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [36]:
# YOUR CODE IS HERE 

ecom3 = ecom.iloc[:, [0, 5, 7]]
ecom3

InvoiceNo  CustomerID  Total
0         536365     17850.0  15.30
1         536365     17850.0  20.34
2         536365     17850.0  22.00
3         536365     17850.0  20.34
4         536365     17850.0  20.34
...          ...         ...    ...
541904    581587     12680.0  10.20
541905    581587     12680.0  12.60
541906    581587     12680.0  16.60
541907    581587     12680.0  16.60
541908    581587     12680.0  14.85

[541909 rows x 3 columns]

Desired Output:

![image.png](https://i.ibb.co/9HsWScR/assignment2k.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 20: Combine "ecom2" and "ecom3" with concat() method</p>

<a id="21"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [37]:
# YOUR CODE IS HERE 

pd.concat([ecom2, ecom3], axis=1)


Total        date InvoiceNo  CustomerID  Total
0       15.30  2010-12-01    536365     17850.0  15.30
1       20.34  2010-12-01    536365     17850.0  20.34
2       22.00  2010-12-01    536365     17850.0  22.00
3       20.34  2010-12-01    536365     17850.0  20.34
4       20.34  2010-12-01    536365     17850.0  20.34
...       ...         ...       ...         ...    ...
541904  10.20  2011-12-09    581587     12680.0  10.20
541905  12.60  2011-12-09    581587     12680.0  12.60
541906  16.60  2011-12-09    581587     12680.0  16.60
541907  16.60  2011-12-09    581587     12680.0  16.60
541908  14.85  2011-12-09    581587     12680.0  14.85

[541909 rows x 5 columns]

Desired Output:

![image.png](https://i.ibb.co/6NGsVMy/assignment2l.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 21: Combine the common (intersected) values of "ecom2" and "ecom3" with merge() method</p>

<a id="22"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

**If "MemoryError" occurs, take only the first 100 rows from both DFs**

In [38]:
# YOUR CODE IS HERE 

ecom_sample= ecom2.head(100)
ecom_sample1 = ecom3.head(100)
merging_df = pd.merge(ecom_sample, ecom_sample1.drop_duplicates(subset=["Total"]), on='Total')
merging_df

Total        date InvoiceNo  CustomerID
0    15.3  2010-12-01    536365     17850.0
1    15.3  2010-12-01    536365     17850.0
2    15.3  2010-12-01    536365     17850.0
3    15.3  2010-12-01    536365     17850.0
4    15.3  2010-12-01    536365     17850.0
..    ...         ...       ...         ...
95   16.5  2010-12-01    536378     14688.0
96   50.4  2010-12-01    536378     14688.0
97   13.2  2010-12-01    536378     14688.0
98   13.2  2010-12-01    536378     14688.0
99   13.2  2010-12-01    536378     14688.0

[100 rows x 4 columns]

Desired Output:

![image.png](https://i.ibb.co/3cgVkjv/assignment2m.png)

In [39]:
ecom2.loc[0:99].merge(ecom3.loc[0:99], on="Total").head(100)

Total        date InvoiceNo  CustomerID
0   15.30  2010-12-01    536365     17850.0
1   15.30  2010-12-01    536373     17850.0
2   15.30  2010-12-01    536373     17850.0
3   15.30  2010-12-01    536375     17850.0
4   15.30  2010-12-01    536375     17850.0
..    ...         ...       ...         ...
95  20.34  2010-12-01    536365     17850.0
96  20.34  2010-12-01    536365     17850.0
97  20.34  2010-12-01    536373     17850.0
98  20.34  2010-12-01    536373     17850.0
99  20.34  2010-12-01    536373     17850.0

[100 rows x 4 columns]

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 22: Combine "ecom2" and "ecom3" with join() method</p>

<a id="23"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [40]:
# YOUR CODE IS HERE 

ecom2.join(ecom3, lsuffix='_left', rsuffix='_right')

Total_left        date InvoiceNo  CustomerID  Total_right
0            15.30  2010-12-01    536365     17850.0        15.30
1            20.34  2010-12-01    536365     17850.0        20.34
2            22.00  2010-12-01    536365     17850.0        22.00
3            20.34  2010-12-01    536365     17850.0        20.34
4            20.34  2010-12-01    536365     17850.0        20.34
...            ...         ...       ...         ...          ...
541904       10.20  2011-12-09    581587     12680.0        10.20
541905       12.60  2011-12-09    581587     12680.0        12.60
541906       16.60  2011-12-09    581587     12680.0        16.60
541907       16.60  2011-12-09    581587     12680.0        16.60
541908       14.85  2011-12-09    581587     12680.0        14.85

[541909 rows x 5 columns]

Desired Output:

![image.png](https://i.ibb.co/DfS2KqQ/assignment2n.png)

## <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">THE END OF THE ASSIGNMENT 02</p>

# Now. it's time to relax!

![Image_Assignment](https://i.ibb.co/jZD3Mqb/relaxing.jpg)

<p style="text-align: center;"><img src="https://docs.google.com/uc?id=1lY0Uj5R04yMY3-ZppPWxqCr5pvBLYPnV" class="img-fluid" 
alt="CLRSWY"></p>

## <p style="background-color:#FDFEFE; font-family:newtimeroman; color:#9d4f8c; font-size:100%; text-align:center; border-radius:10px 10px;">WAY TO REINVENT YOURSELF</p>

<a id="24"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>
___